### PR TITLE
[Frontend] Add leaf constructs and association to OpenMP/ACC directives

### PR DESCRIPTION
**Pull request by kparzysz**
Add members "leafConstructs" and "association" to .td describing OpenMP/ACC directives. The naming follows the terminology used in the OpenMP standard: a "leaf" construct is a construct that is itself not a composition or a combination of other constructs, and "association" is the source language construct to which the directive applies (e.g. loop, block, etc.)

The tblgen-generated output then contains two additional functions
- getLeafConstructs(D), and
- getDirectiveAssociation(D) plus "enum class Association", all in namespaces "llvm::omp" and "llvm::acc".

Note: getLeafConstructs returns an empty sequence for a construct that is itself a leaf construct.

Use the new functions to simplify a few OpenMP-related functions in clang.

### CODE REVIEW

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (alexey-bataev) [inline clang/lib/Basic/OpenMPKinds.cpp:624]: Leafs

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (alexey-bataev) [inline clang/lib/Basic/OpenMPKinds.cpp:635]: There is no OMPD_end_do_simd in clang

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (alexey-bataev) [inline clang/lib/Basic/OpenMPKinds.cpp:646]: Leafs

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (alexey-bataev) [inline clang/lib/Basic/OpenMPKinds.cpp:658]: Leafs

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: Done

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: Done

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: Done

**Comment by kparzysz** [comment]
I removed the check for `end do simd` with a comment that it can't appear in clang.  I think you deleted that remark, let me know if I should bring the check back.

**Review by clementval — APPROVED** [review]
Looks good to me from the OpenACC point of view. Someone from OpenMP should approve as well.

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (erichkeane) [reply clang/lib/Basic/OpenMPKinds.cpp]: Plural of 'leaf' is 'leaves'.  'leafs' is a canadian hockey team :D

Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: Leafs is an accepted form.  I used to live in Toronto, btw.  :)

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: Removed.

**Comment by kparzysz** [comment]
The failing test on Windows is unrelated.  I've seen it fail in other builds before.

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (kiranchandramohan) [inline clang/lib/Basic/OpenMPKinds.cpp:604]: Nit: A comment that workshare is not applicable to C/C++ might be helpful.

**Review by kiranchandramohan — APPROVED** [review]
> Add members "leafs" and "association" to .td describing OpenMP/ACC directives: "leafs" are the leaf constructs for composite/combined constructs, and "association" is the source language construct to which the directive applies (e.g. loop, block, etc.)

Could you clarify that these are terminology used in the OpenMP standard?

Great work @kparzysz.

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp:604]: I removed the check.  Formally, it does fall under the "parallel" family of directives, so I think we shouldn't exclude it on the basis that it will never show up.

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (alexey-bataev) [inline clang/lib/Basic/OpenMPKinds.cpp:622]: Expand auto here and in other places (ArrayRef<Directive>)

File: llvm/test/TableGen/directive1.td
Comment (alexey-bataev) [inline llvm/test/TableGen/directive1.td:359]:
```suggestion
// IMPL-NEXT:  ArrayRef<llvm::tdl::Directive> llvm::tdl::getLeafConstructs(llvm::tdl::Directive Dir) {
```

File: llvm/utils/TableGen/DirectiveEmitter.cpp
Comment (alexey-bataev) [inline llvm/utils/TableGen/DirectiveEmitter.cpp:247]:
```suggestion
  OS << "ArrayRef<Directive> getLeafConstructs(Directive D);\n";
```

Comment (kparzysz) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:247]: I don't think this is a good idea.  Returning a reference is pretty much free, Returning an ArrayRef will always create a temporary object.  You can always use ArrayRef in the caller, if that's what's desired there.

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: Done

File: llvm/utils/TableGen/DirectiveEmitter.cpp
Comment (alexey-bataev) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:247]: No, it won't. ArrayRef is not an owning container, so it does not create a copy.

Comment (kparzysz) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:247]: The ArrayRef will be the temporary object.

Comment (alexey-bataev) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:247]: No, it is not.

Comment (kparzysz) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:247]: I'm expecting you to be more professional in your replies.  I'm simply trying to understand your motivation here.

Comment (alexey-bataev) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:247]: Read the documentation https://llvm.org/docs/ProgrammersManual.html#llvm-adt-arrayref-h

File: llvm/utils/TableGen/DirectiveEmitter.cpp
Comment (alexey-bataev) [inline llvm/utils/TableGen/DirectiveEmitter.cpp:465]: Looks like this approach breaks https://llvm.org/docs/CodingStandards.html#do-not-use-static-constructors. Better to avoid static SmallVectors (IIUC you're creating static SmallVector vars here) and use POD instead to avoid use of static constructors/destructors.

**Review by alexey-bataev — CHANGES_REQUESTED** [review]

File: llvm/utils/TableGen/DirectiveEmitter.cpp
Comment (kparzysz) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:465]: Sure, will do.

File: llvm/utils/TableGen/DirectiveEmitter.cpp
Comment (kparzysz) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:247]: That's about input parameters (because then we don't restrict what objects can be passed, i.e. converted to an ArrayRef).  For return values it doesn't matter, at least from the point of view of interface flexibility.  The static constructor issue you pointed out will need this though.

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (skatrak) [reply clang/lib/Basic/OpenMPKinds.cpp]: > Leafs is an accepted form.

Could you point to some source where that's described? After a quick search, I find that "leaves" is the widely accepted plural of "leaf", and that "leafs" is only recognized by certain dictionaries. One instance where it's not: https://dictionary.cambridge.org/dictionary/english/leaf.

Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: Merriam-Webster is more oriented towards American English.  The [entry for leaf](https://www.merriam-webster.com/dictionary/leaf) shows "leafs" as an alternative plural form.

File: llvm/utils/TableGen/DirectiveEmitter.cpp
Comment (kparzysz) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:465]: Done.

File: clang/lib/Basic/OpenMPKinds.cpp
Comment (erichkeane) [reply clang/lib/Basic/OpenMPKinds.cpp]: Huh, interestingly, that Merriam Webster link is the only one I can find that acknowledges `leafs` as a plural for the general case.  The only other case I could find was wiktionary: https://en.wiktionary.org/wiki/leaf

Which shows that `leafs` is valid in 2 cases: when talking about multi-section doors, or when used as a pejorative for a Canadian.  So I don't really know anything anymore.

That said, while LLVM's policy is only 'english' and doesn't specify American, Canadian, or British English, by practice we use American English spellings for things.

Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: If you strongly prefer the more conformist spelling, I can change it...  :)

Comment (erichkeane) [reply clang/lib/Basic/OpenMPKinds.cpp]: 'strongly prefer'?  No.  Mildly? Yeah, mildly.  Perhaps slightly more than mildly because it is in an API and I'd prefer we be consistent with the 'more common' spelling.

Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: Uh, the "leafs" is only used as a local variable name, the function name has singular "Leaf".

Comment (kparzysz) [reply clang/lib/Basic/OpenMPKinds.cpp]: I've changed it.

**Comment by kparzysz** [comment]
Hey @alexey-bataev, I have made all the changes you requested.  Could you take another look?

File: llvm/test/TableGen/directive1.td
Comment (alexey-bataev) [inline llvm/test/TableGen/directive1.td:363]: Is this supposed to be a container for the leaves? ArrayRef is not owning container, I'm afraid it may cause some UB. Better just to use POD arrays here, if I understand correctly generated code.

Comment (kparzysz) [reply llvm/test/TableGen/directive1.td:363]: Eh, sorry, it's a mistake.  Thanks for the catch.

Comment (kparzysz) [reply llvm/test/TableGen/directive1.td]: Returning an empty ArrayRef by value now.  I can't use regular array (as in the case of non-empty sets), because I can't create a zero-length array.  The empty ArrayRef does not refer to any storage, so returning it by value should be ok.

File: llvm/utils/TableGen/DirectiveEmitter.cpp
Comment (alexey-bataev) [inline llvm/utils/TableGen/DirectiveEmitter.cpp:504]: The actual storage here is just POD arrays, right?

Comment (kparzysz) [reply llvm/utils/TableGen/DirectiveEmitter.cpp:504]: Yes,
```
static const llvm::omp::Directive[] = {...};
```

**Comment by kparzysz** [comment]
Hey @erichkeane, do you have anything to add?  I have replaced `leafs` in the .td file with `leafConstructs`, and with `leaves` elsewhere.

**Review by alexey-bataev — APPROVED** [review]
LG

**Comment by erichkeane** [comment]
> Hey @erichkeane, do you have anything to add? I have replaced `leafs` in the .td file with `leafConstructs`, and with `leaves` elsewhere.

Nothing left, mine was just a drive-by comment that got out of hand :)  LGTM.

**Pull request merged**
